### PR TITLE
fix: do not skip graceful preemption for tasks using alternate rendezvous path [DET-9258]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Determined on HPC clusters used an alternate path for rendezvous, but our preemption logic already checked if rendezvous on the traditional path was ready to determine if something should be gracefully preempted or killed, causing HPC jobs to always get killed.

## Test Plan
- [x] test on casablanca
`det e pause` doesn't involve slurm in the termination at all, as it should work
```
[2023-04-04T20:16:20.520675Z] [cont=0] || INFO: [3734213] root: Validating checkpoint storage ...
[2023-04-04T20:16:20.531478Z] [cont=0] || INFO: [3734213] root: Launching: ['python3', '-m', 'determined.launch.horovod', '--autohorovod', '--trial', 'model_def:MNistTrial']
[2023-04-04T20:16:31.758060Z]          || INFO: gracefully terminating allocation's remaining resources (reason: user requested pause)
[2023-04-04T20:16:33.649788Z] [cont=0] || INFO: [3734217] root: Creating _PyTorchTrialController with MNistTrial.
[2023-04-04T20:16:34.851242Z] [cont=0] || INFO: [3734217] root: Downloading https://s3-us-west-2.amazonaws.com/determined-ai-test-data/pytorch_mnist.tar.gz to /tmp/data-rank0/MNIST/pytorch_mnist.tar.gz
[2023-04-04T20:17:00.664254Z] [cont=0] || /opt/conda/lib/python3.8/site-packages/torch/nn/functional.py:1331: UserWarning: dropout2d: Received a 2-D input to dropout2d, which is deprecated and will result in an error in a future release. To retain the behavior and silence this warning, please use dropout instead. Note that dropout2d exists to provide channel-wise dropout on inputs with 2 spatial dimensions, a channel dimension, and an optional batch dimension (i.e. 3D or 4D inputs).
[2023-04-04T20:17:00.664325Z] [cont=0] ||   warnings.warn(warn_msg)
[2023-04-04T20:17:00.664881Z] [cont=0] || /opt/conda/lib/python3.8/site-packages/torch/nn/modules/container.py:139: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
[2023-04-04T20:17:00.664914Z] [cont=0] ||   input = module(input)
[2023-04-04T20:17:07.325324Z] [cont=0] || INFO: [3734217] determined.core: report_training_metrics(steps_completed=100, metrics={'loss': 0.6069826483726501})
[2023-04-04T20:17:13.866477Z] [cont=0] || INFO: [3734217] determined.core: Reported checkpoint to master ab8141c7-d24d-473d-92d4-6f9fa6350c31
[2023-04-04T20:17:13.866582Z] [cont=0] || INFO: [3734217] root: Waiting for Tensorboard files to finish uploading
[2023-04-04T20:17:26.804712Z]          || INFO: resources exited successfully with a zero exit code
[2023-04-04T20:17:26.848108Z]          || INFO: Trial (Experiment 10) was terminated: allocation stopped after resources exited successfully with a zero exit code
```
- [x] look for similar bugs

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/task/allocation.go [hunk 1]
```diff
@@ -1159,6 +1162,16 @@ func (a *Allocation) getModelState() model.AllocationState {
 	return *a.model.State
 }
 
+func (a *Allocation) ready() bool {
+	// Most trials use `a.rendezvous` and the normal rendezvous APIs, and go through this path.
+	return (a.rendezvous != nil && a.rendezvous.ready()) ||
```
Comment (jerryharrow) [inline master/internal/task/allocation.go:1167]: So when is rendezvous used?   I do not see it triggered with 2 node dtrain job (cifar10_pytorch).

Comment (stoksc) [reply master/internal/task/allocation.go:1167]: it is used in the non-hpc mode. it is https://github.com/determined-ai/determined/blob/master/harness/determined/exec/prep_container.py#L69 vs https://github.com/determined-ai/determined/blob/master/harness/determined/exec/prep_container.py#L82

**Review by jerryharrow — APPROVED** [review]
lgtm

**Comment by jerryharrow** [comment]
I think this might have broken HPC job cancellation

https://hpe-aiatscale.atlassian.net/browse/DET-9284

**Pull request merged**